server/src/main/java/io/crate/metadata/IndexReference.java

```diff
@@ -110,7 +109,14 @@ public IndexReference build(Map<ColumnIdent, Reference> references) {
                 // This code handles outdated shards case.
                 return new IndexReference(position, oid, isDropped, ident, indexType, columns, analyzer);
             }
-            List<Reference> sources = references.values().stream().filter(ref -> sourceNames.contains(ref.storageIdent())).collect(Collectors.toList());
+            List<Reference> sources = new ArrayList<>(sourceNames.size());
+            for (String sourceName : sourceNames) {
+                Reference ref = references.values().stream()
+                    .filter(r -> r.storageIdent().equals(sourceName))
+                    .findAny()
+                    .orElseThrow();
+                sources.add(ref);
+            }
```

Comment:
Before this relied on the ordering of `references` - which was a LinkedHashSet, ordered by position and name. 
The change here preserves the ordering of the source definition. E.g. `index index_ft using fulltext("col_e", "col_f")` will result in the order `col_e` > `col_f`, no matter their position

It also raises an error if the column is missing, because then there's clearly something wrong.